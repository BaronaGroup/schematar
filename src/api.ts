import fs from 'fs'
import path from 'path'

import mkdirp from 'mkdirp'

import ComplexType from './complex'
import { generateHash } from './hash-schema'
import jsonSchemaOutput, { JSONSchemaOptions } from './json-schema-output'
import { karhu } from './karhu'
import mongooseOutput from './mongoose-output'
import nowToken from './now'
import { ObjectId as ObjectIdType } from './object-id'
import { Schema as SchemaType } from './schema'
import tsOutput, { TSOptions } from './ts-output'

export const ObjectId = ObjectIdType
export type Schema = SchemaType
export const Complex = ComplexType
export const now = nowToken
const log = karhu('schematar-api')

export function createMongooseSchema(schema: Schema, context = 'mongoose') {
  return mongooseOutput(schema, context)
}

export function createTypescriptInterfaceDefinition(
  exportName: string,
  schema: Schema | ComplexType,
  context: string = 'typescript',
  options: TSOptions = {}
) {
  return tsOutput(exportName, schema, context, options)
}

export function createJSONSchema(schema: Schema, context: string = 'jsonschema', options: JSONSchemaOptions = {}) {
  return jsonSchemaOutput(schema, context, options)
}

export async function createTypescriptInterfaceFiles(
  sourceFileGlobOrFileArray: string | string[],
  outputPath: string,
  logCreations = false
) {
  const { default: glob } = await import('glob')
  await new Promise((resolve, reject) => mkdirp(outputPath, (err) => (!err ? resolve() : reject(err))))
  const files = [].concat(
    ...(Array.isArray(sourceFileGlobOrFileArray)
      ? sourceFileGlobOrFileArray.map((f) => glob.sync(f))
      : [glob.sync(sourceFileGlobOrFileArray)])
  )
  for (const file of files) {
    await createTSInterfaceFile(file, outputPath, logCreations)
  }
}

export interface TypescriptSchemaDefinition {
  name: string
  context?: string
  omitExtraExports?: boolean
  schema?: Schema | ComplexType
  exportHash?: string
  doNotImportObjectId?: boolean
}

interface SchemaFile {
  name?: string
  default?: Schema
  schema?: Schema
  typescriptSchemas?: TypescriptSchemaDefinition[]
}

export async function createTSInterfaceFile(
  filename: string,
  outputPath: string,
  logCreations: boolean = true,
  formatOutputFilename = defaultOutputFilenameFormatter
) {
  const schemaFile = require(path.resolve(filename)) as SchemaFile
  const topLevelSchema = schemaFile.schema || schemaFile.default
  const topLevelName = schemaFile.name || pickNameFromFilename(filename)
  const exported: string[] = []
  if (schemaFile.typescriptSchemas) {
    const typescriptSchemas = schemaFile.typescriptSchemas
    for (const schemaDef of typescriptSchemas) {
      const schema = schemaDef.schema || topLevelSchema,
        name = schemaDef.name || topLevelName

      if (!schema) throw new Error('No schema found for ' + name)
      exported.push(
        createTypescriptInterfaceDefinition(name, schema, schemaDef.context || 'typescript', {
          ...schemaDef,
          doNotImportObjectId: schemaDef.doNotImportObjectId || typescriptSchemas.indexOf(schemaDef) > 0,
        })
      )
    }
  } else if (topLevelSchema) {
    exported.push(createTypescriptInterfaceDefinition(topLevelName, topLevelSchema))
  }
  const outputFilename = path.join(outputPath, formatOutputFilename(filename))

  if (logCreations) {
    log.info('Storing typescript interfaces in', outputFilename)
  }

  if (exported.length) {
    await new Promise((resolve, reject) =>
      fs.writeFile(outputFilename, exported.join('\n'), 'utf8', (err) => (!err ? resolve() : reject(err)))
    )
  }
}

function defaultOutputFilenameFormatter(inputFilename: string) {
  const ext = path.extname(inputFilename),
    basename = path.basename(inputFilename, ext)

  return basename + '.schematar.ts'
}

function pickNameFromFilename(filename: string) {
  const fn = path.basename(filename, path.extname(filename))
  return (fn[0].toUpperCase() + fn.substring(1))
    .replace(/[\-_][a-z]/g, (match) => match[1].toUpperCase())
    .replace(/[^a-zA-Z0-9]/g, '')
}

export function hashSchema(schema: Schema | ComplexType) {
  return generateHash(schema)
}
