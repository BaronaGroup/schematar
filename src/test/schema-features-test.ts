import {Schema} from '../api'
import {testJSONSchema, testMongooseField, testTypescriptInterface} from './test-utils'

describe('schema-features-test', function () {
  describe('optional', function () {
    const schema: Schema = {
      fields: {
        field: {type: Number, optional: true}
      }
    }

    it('json-schema', testJSONSchema(schema, data => {
      expect(data).toMatchSnapshot()
    }))
/*  Not supported in mongoose; everything is optional
    it('mongoose', testMongooseField(schema, 'field', field => {
      expect(field).toEqual({type: Number})
    }))*/

    it('typescript', testTypescriptInterface(schema, tsInterface => {
      expect(tsInterface).toMatchSnapshot()
    }))
  })
  describe('optionalIn', function () {
    const schema: Schema = {
      fields: {
        field: {type: Number, optionalIn: ['o1', 'o2', 'o3']}
      }
    }

    describe('match', function() {
      it('json-schema', testJSONSchema(schema, data => {
        expect(data).toMatchSnapshot()
      }, {}, 'o2'))

      it('typescript', testTypescriptInterface(schema, tsInterface => {
        expect(tsInterface).toMatchSnapshot()
      }, {}, 'o2'))
    })

    describe('no match', function() {
      it('json-schema', testJSONSchema(schema, data => {
        expect(data).toMatchSnapshot()
      }))

      it('typescript', testTypescriptInterface(schema, tsInterface => {
        expect(tsInterface).toMatchSnapshot()
      }))
    })
  })

  describe('presentIn', function () {
    const schema: Schema = {
      fields: {
        field: {type: Number, presentIn: ['p1', 'p2', 'p3']}
      }
    }

    describe('match', function() {
      it('json-schema', testJSONSchema(schema, data => {
        expect(data).toMatchSnapshot()
      }, {}, 'p2'))

      it('typescript', testTypescriptInterface(schema, tsInterface => {
        expect(tsInterface).toMatchSnapshot()
      }, {}, 'p2'))

      it('mongoose', testMongooseField(schema, 'field', result => {
        expect(result).toEqual({type: Number})
      }, 'p2'))
    })

    describe('no match', function() {
      it('json-schema', testJSONSchema(schema, data => {
        expect(data).toMatchSnapshot()
      }))

      it('typescript', testTypescriptInterface(schema, tsInterface => {
        expect(tsInterface).toMatchSnapshot()
      }))

      it('mongoose', testMongooseField(schema, 'field', result => {
        expect(result).toBeUndefined()
      }))
    })
  })
})
