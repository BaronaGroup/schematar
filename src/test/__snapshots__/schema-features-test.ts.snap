// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema-features-test optional json-schema 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"properties\\": {
    \\"field\\": {
      \\"type\\": \\"number\\"
    }
  },
  \\"required\\": []
}"
`;

exports[`schema-features-test optional typescript 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface TestInterfaceBase<IDType, DateType> {
  field?: number
}"
`;

exports[`schema-features-test optionalIn match json-schema 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"properties\\": {
    \\"field\\": {
      \\"type\\": \\"number\\"
    }
  },
  \\"required\\": []
}"
`;

exports[`schema-features-test optionalIn match typescript 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface TestInterfaceBase<IDType, DateType> {
  field?: number
}
import {ObjectId} from 'mongodb'
export type TestInterfaceMongoose = TestInterfaceBase<ObjectId, Date>
export type TestInterfaceJSON = TestInterfaceBase<string, string>
export type TestInterfaceFluid = TestInterfaceBase<string | ObjectId, string | Date>"
`;

exports[`schema-features-test optionalIn no match json-schema 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"properties\\": {
    \\"field\\": {
      \\"type\\": \\"number\\"
    }
  },
  \\"required\\": [
    \\"field\\"
  ]
}"
`;

exports[`schema-features-test optionalIn no match typescript 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface TestInterfaceBase<IDType, DateType> {
  field: number
}"
`;

exports[`schema-features-test presentIn match json-schema 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"properties\\": {
    \\"field\\": {
      \\"type\\": \\"number\\"
    }
  },
  \\"required\\": [
    \\"field\\"
  ]
}"
`;

exports[`schema-features-test presentIn match typescript 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface TestInterfaceBase<IDType, DateType> {
  field: number
}
import {ObjectId} from 'mongodb'
export type TestInterfaceMongoose = TestInterfaceBase<ObjectId, Date>
export type TestInterfaceJSON = TestInterfaceBase<string, string>
export type TestInterfaceFluid = TestInterfaceBase<string | ObjectId, string | Date>"
`;

exports[`schema-features-test presentIn no match json-schema 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"properties\\": {},
  \\"required\\": []
}"
`;

exports[`schema-features-test presentIn no match typescript 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface TestInterfaceBase<IDType, DateType> {
}"
`;
